### PR TITLE
YONK-852: Course Settings added to provide support to set language

### PR DESCRIPTION
**Pull request by aamishbaloch**
Changes include the addition of Course Settings to provide support to set language of a course. Unit tests added also.

### CODE REVIEW

**Comment by aamishbaloch** [comment]
@ziafazal Can you review this?

File: course_metadata/tests/tests.py [hunk 1]
```diff
@@ -77,3 +78,53 @@ def test_get_course_aggregate_metadata_by_course_key(self):
...
+
+
+class CourseSettingTests(ModuleStoreTestCase, APIClientMixin):
+    """ Test suite for Course Setting """
+
+    MODULESTORE = TEST_DATA_SPLIT_MODULESTORE
+
+    def setUp(self):
+        super(CourseSettingTests, self).setUp()
+        self.base_course_metadata_uri = '/api/server/courses_metadata/'
```
Comment (ziafazal) [inline course_metadata/tests/tests.py]: Can use `reverse` method to get url of the course setting view?

File: course_metadata/tests/tests.py [hunk 2]
```diff
@@ -77,3 +78,53 @@ def test_get_course_aggregate_metadata_by_course_key(self):
...
+        """
+        data = {
+            "languages": self.languages
+        }
+
+        response = self.do_post(
+            "{}{}{}".format(self.base_course_metadata_uri, self.course.id, self.settings_uri),
+            data,
+        )
+        self.assertEqual(response.status_code, 200)
```
Comment (ziafazal) [inline course_metadata/tests/tests.py]: Should not the response code here be 201?

File: course_metadata/tests/tests.py [hunk 3]
```diff
@@ -77,3 +78,53 @@ def test_get_course_aggregate_metadata_by_course_key(self):
...
+
+        course_settings = CourseSetting.objects.get(id=self.course.id)
```
Comment (ziafazal) [inline course_metadata/tests/tests.py]: I think we should verify creation of course setting via GET call.

File: course_metadata/tests/tests.py [hunk 4]
```diff
@@ -77,3 +78,53 @@ def test_get_course_aggregate_metadata_by_course_key(self):
...
+        languages = self.languages.lower().replace(" ", "").split(',')
+        for language in course_settings.languages_list:
+            self.assertIn(language, languages)
+
+    def test_course_settings_with_no_languages(self):
```
Comment (ziafazal) [inline course_metadata/tests/tests.py]: We should add a unit test for updating course languages.

File: course_metadata/tests/tests.py [hunk 5]
```diff
@@ -77,3 +78,53 @@ def test_get_course_aggregate_metadata_by_course_key(self):
...
+        """
+        Test for testing course languages in course settings
+        """
+        data = {
+            "languages": "",
+        }
+
+        response = self.do_post(
+            "{}{}{}".format(self.base_course_metadata_uri, self.course.id, self.settings_uri),
```
Comment (ziafazal) [inline course_metadata/tests/tests.py]: can we use `self.settings_uri` here?.

File: course_metadata/tests/tests.py [hunk 6]
```diff
@@ -77,3 +78,53 @@ def test_get_course_aggregate_metadata_by_course_key(self):
...
+            data,
+        )
+        self.assertEqual(response.status_code, 200)
+
+        course_settings = CourseSetting.objects.get(id=self.course.id)
```
Comment (ziafazal) [inline course_metadata/tests/tests.py]: can we verify course setting via GET call here?

File: course_metadata/urls.py [hunk 7]
```diff
@@ -0,0 +1,11 @@
+from django.conf import settings
+from django.conf.urls import patterns, url
+
+from . import views
+
+COURSE_ID_PATTERN = settings.COURSE_ID_PATTERN
+
+urlpatterns = patterns(
+    '',
+    url(r'^{}/settings$'.format(COURSE_ID_PATTERN), views.CourseSettingView.as_view(), name='course_settings'),
```
Comment (ziafazal) [inline course_metadata/urls.py]: can we make view name `course_settings` bit unique. may be `additional_course_settings`?

File: course_metadata/views.py [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+from rest_framework import status
+from rest_framework.response import Response
+
+from edx_solutions_api_integration.permissions import SecureAPIView
+from edx_solutions_api_integration.courseware_access import get_course_key
+
+from course_metadata.models import CourseSetting
+
+
+class CourseSettingView(SecureAPIView):
```
Comment (ziafazal) [inline course_metadata/views.py]: Permission of this view should be staff or readonly.

File: course_metadata/views.py [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+
+        **POST**
+
+        If the request is successful, the request returns an HTTP 200 "OK" response.
+    """
+    def post(self, request, course_id):
+        course_key = get_course_key(course_id)
+        languages = request.data.get('languages')
+
+        try:
```
Comment (ziafazal) [inline course_metadata/views.py]: Can we use `CourseSettingSerializer ` here?

**Review by ziafazal — CHANGES_REQUESTED** [review]
@aamishbaloch completed first round of review.

**Comment by aamishbaloch** [comment]
@ziafazal Can you review the changes?

File: course_metadata/serializers.py [hunk 10]
```diff
@@ -0,0 +1,15 @@
...
+from course_metadata.models import CourseSetting
+
+
+class CourseSettingSerializer(serializers.ModelSerializer):
+    id = CourseKeyField(required=False)
+
+    class Meta:
+        model = CourseSetting
+
+    def to_representation(self, instance):
```
Comment (ziafazal) [inline course_metadata/serializers.py]: can we create a field to return list of languages?

**Review by ziafazal — CHANGES_REQUESTED** [review]
@aamishbaloch can you apply staff and readonly permissions and set languages to ListField in serializer?

**Comment by aamishbaloch** [comment]
@ziafazal Can you review the changes?

File: course_metadata/tests/tests.py [hunk 11]
```diff
@@ -77,3 +81,63 @@ def test_get_course_aggregate_metadata_by_course_key(self):
...
+        Test for adding course languages in course settings
+        """
+        data = {"languages": self.languages}
+        response = self.do_put(self.course_settings_uri, data)
+
+        self.assertEqual(response.status_code, 200)
+        for language in response.data['languages']:
+            self.assertIn(language, self.languages)
+
+    def test_course_settings_get_languages(self):
```
Comment (ziafazal) [inline course_metadata/tests/tests.py]: This test seems redundant.

File: course_metadata/tests/tests.py [hunk 12]
```diff
@@ -77,3 +81,63 @@ def test_get_course_aggregate_metadata_by_course_key(self):
...
+        data = {"languages": self.languages}
+        response = self.do_put(self.course_settings_uri, data)
+        self.assertEqual(response.status_code, 200)
+
+        response = self.do_get(self.course_settings_uri)
+        self.assertEqual(response.status_code, 200)
+        for language in response.data['languages']:
+            self.assertIn(language, self.languages)
+
+    def test_course_settings_update_languages(self):
```
Comment (ziafazal) [inline course_metadata/tests/tests.py:109]: can we add a test case for non-staff user permission validation?

**Review by ziafazal — CHANGES_REQUESTED** [review]
@aamishbaloch couple of minor test case related suggestions.

**Review by ziafazal — APPROVED** [review]
👍

**Pull request merged**